File: ruby/10_regular_expression_matching.rb
# @param {String} s
# @param {String} p
# @return {Boolean}

# 判断字符串是否匹配
# p 为模式串，含有 `*` , `.`两种特殊字符
# . 可以表示任何一个 a-z 的字符
# * 可以表示0个或多个*前面的字符


# 该方法使用递归
# 1.首先判断p长度如果 `>=` 2， 则看第二个是否为*（第一个字符为*其实可以当作普通字符来看待）。
# 2.如果第二个字符不是*，则依次判断第一个字符是否匹配，然后递归的进行判断第二个字符；
# 3.如果第二个字符是*，则分两种情况：
# 3.1 可能存在*代表0个前面的字符，所以直接判断s和p从idx为2开始的子串；
# 3.2 也有可能存在*代表1个或多个前面的字符，所以先判断第一个字符是否匹配，然后递归的判断s从idx为1的子串和p；


def is_match(s, p)

	return is_match(s, 0, p, 0)

	if p.empty?
		return s.empty?
	end

	first_match = s.length > 0 && (s[0] == p[0] || p[0] == '.')

	if p.length >= 2 && p[1] == '*'
		return is_match(s, p[2..p.length-1]) || ( first_match && is_match(s[1..s.length-1], p) )
	else
		return first_match && is_match( s[1..s.length-1], p[1..p.length-1] )
	end
end


def is_match(s, sBegin, p, pBegin)

	if pBegin >= p.length
		return sBegin >= s.length
	end

	first_match = sBegin < s.length && ( s[sBegin] == p[pBegin] || p[pBegin] == '.' )

	if pBegin <= p.length - 2 && p[pBegin + 1] == '*'
		return is_match(s, sBegin, p, pBegin + 2) || (first_match && is_match(s, sBegin+1, p, pBegin))
	else
		return first_match && is_match(s, sBegin+1, p ,pBegin+1)
	end

end

File: cpp/10_regular_expression_matching.cpp
#include <iostream>
#include <vector>
#include <string>
#include <stack>
#include <queue>
#include <map>
#include <set>
#include <unordered_set>
#include <unordered_map>
using namespace std;

// p含有`.`, `*`，分别表示任意一个字符，和任意0个或多个前面的字符。判断s和p是否匹配。
class Solution {
public:
    bool isMatch(string s, string p) {
		if (p.empty()) { return s.empty(); }

		bool first_match = s.size() > 1 && (p[0]==s[0] || p[0]=='.');

		if (p.size() == 1) { return s.size() == 1 && first_match; }

		if (p[1] != '*') {
			return first_match && isMatch(s.substr(1), p.substr(1));
		} else {
			bool one = isMatch(s, p.substr(2));
			bool two = first_match && isMatch(s.substr(1), p);
			return one || two;
		}
    }

	bool isMatch_dp(string s, string p) {
		int m = (int)s.size(), n = (int)p.size();
		vector<vector<bool>> dp(m+1, vector(n+1, false));
		dp[0][0] = true;

		for (int i = 1; i <= n; ++i) {
			dp[0][i] = (i>1) && (p[i-1] == '*') && dp[0][i-2];
		}

		for (int i = 1; i <=n; ++i) {
			for (int j = 1; j <=m; ++j) {
				if (p[i-1] == '*') {
					bool one = dp[j-1][i] && (s[j-1]==p[i-2] || p[i-2]=='.');
					bool two = (i-2 >= 0) && dp[j][i-2]; // case s="mi", p = "mis*"
					dp[j][i] = one || two;
				} else {
					dp[j][i] = (p[i-1]==s[j-1] || p[i-1]=='.') && dp[j-1][i-1]; 
				}
			}
		}

		return dp[m][n];
	}
};
